### ISSUE TITLE
Replace PhantomJS with headless Chrome

### ISSUE THREAD

**Issue by mkllnk** [openfoodfoundation/openfoodnetwork#2469]
We see a lot of test failures on our Continuous Integration servers. Some of them could be related to PhantomJS not being very consistent and reliable.

## Description

Rob has been working on using headless Chrome instead of PhantomJS for our feature tests. I thought that was a really good idea, but it is actually a lot of work to change our tests to work with Chrome. I'm also wondering if Chrome is what we want. In an ideal world, our tests would run with all browsers, but at the moment the tests don't pass in all browsers (different test behaviour) and we don't have enough CI power.

So assuming that we will test with only one browser, here is a little breakdown:

For PhantomJS:
- It's like an old Safari. We test compatibility with older browsers.
- It's made for testing and can raise Javascript errors as Ruby errors.
- We are used to it and we don't have to change tests.

For Chrome:
- PhantomJS is not maintained any more: https://github.com/ariya/phantomjs/issues/15344
- Chrome is more reliable.
- It behaves like a real browser, not something emulating a browser.
- Chrome has a global market share of ~50%. https://en.wikipedia.org/wiki/Usage_share_of_web_browsers
- A simple option changed on a developer's machine can activate real browsing to see what's going on (not headless).

I tend to switch to Chrome, but I wanted to double check with you if you have any concerns.

## Expected Behavior

Tests always pass.

## Actual Behavior

Tests sometimes pass.

## Steps to Reproduce

Open a pull request to trigger a test run.

## Context

When developing, our test servers often have different results to my own machine. And I can't really see what PhantomJS is doing apart from `save_and_create_screenshot`.

## Severity

s4 tech debt

## Possible Fix

Use headless Chrome or Firefox. Rob started that work and I added a little bit. Now we are managing a branch as a team, creating pull requests to fix the broken tests: https://github.com/openfoodfoundation/openfoodnetwork/compare/headless-chrome?expand=1

**7** Broken tests:
```
rspec ./spec/features/admin/variant_overrides_spec.rb:289 -> #3326

rspec ./spec/features/admin/bulk_order_management_spec.rb:485 -> #3325
rspec ./spec/features/admin/shipping_methods_spec.rb:79
rspec ./spec/features/admin/product_import_spec.rb:130

rspec ./spec/features/admin/customers_spec.rb:24
rspec ./spec/features/admin/enterprises/images_spec.rb:29
rspec ./spec/features/admin/enterprises/images_spec.rb:66
```

**Comment by Matt-Yorkley** [comment]
The issue with PhantomJS is that we're effectively testing our site with a broken, outdated version of a browser that nobody uses, and expecting the results to be useful. It's not great for integration tests if the thing the tests are interacting with doesn't even look like our site. Here's a visual example from Gitlab:

PhantomJS:
![phantomjs](https://about.gitlab.com/images/blogimages/moving-to-headless-chrome/render-phantomjs.png)

Headless Chrome:
![chromeless](https://about.gitlab.com/images/blogimages/moving-to-headless-chrome/render-chrome.png)

**Comment by sauloperez** [comment]
I see, it's like having no tests at all

**Comment by sauloperez** [comment]
Now that we are at it, this getting urgent IMO. Is there anyone with room and willing to take this one soon? This would bring some reliability to our flaky features specs, isn't it?

**Comment by luisramos0** [comment]
the spree upgrade is almost over and I need another broken build to fix 🤣 
seriously, this is something I'd be happy to help with, that's why I am asking for a branch.

**Comment by mkllnk** [comment]
@luisramos0 Sure, I pushed the branch to the main repo. Main work has been done by Rob before me.

About the process: There is a lot of WIP in that branch. I think it needs some history rewriting to make it easier to understand and easier to rebase. So I would suggest a modification to the process:

- Most contributions are done via pull requests to that branch, same as Spree upgrade.
- But there is one person at a time who has the right to rebase/rewrite the branch to clean it up.
- We may decide to backport some of the changes into the master branch and then remove them from the history in the upgrade branch.

It will also make it easier to review the whole branch in the end. That is something we can't do with the Spree upgrade branch any more.

**Comment by luisramos0** [comment]
awesome, new branch! thanks @mkllnk 
https://github.com/openfoodfoundation/openfoodnetwork/tree/headless-chrome

This is what I have done with it so far:
```
git checkout master
git pull upstream master
git checkout headless-chrome
git reset --hard upstream/headless-chrome
git rebase master
git add spec/features/consumer/multilingual_spec.rb
git rebase --continue
git add spec/features/consumer/multilingual_spec.rb
git rebase --continue
git push -f upstream headless-chrome
```

After this I moved 3 "travis wip" commits to a separate branch and PR, see #3086
We can decide what to do with these ones on the PR.

I fixed a commit message, new message is "Update checkout_workflow and spec_helper to work with Selenium".

Looking at the semaphore build, there are "only" 12 broken tests.

Does all this sounds correct Maikel? Sounds too easy!
This spree upgrade makes everything look easy 🤣

**Comment by sauloperez** [comment]
only 12?? maybe it's wasn't that crazy after all

**Comment by luisramos0** [comment]
Actually it's 10 broken tests.

New PR #3089 fixes 1 test:
- [ ] rspec/features/consumer/shopping/embedded_shopfronts_spec.rb:46
New PR #3092 fixes another one:
- [ ] spec/features/admin/subscriptions_spec.rb:27

So, we have 8 remaining broken tests to handle now:
- [ ] rspec ./spec/features/admin/customers_spec.rb:24
- [ ] spec/features/admin/bulk_order_management_spec.rb:450 (this one only fails sometimes)
- [ ] rspec ./spec/features/admin/bulk_order_management_spec.rb:485 
- [ ] rspec ./spec/features/admin/shipping_methods_spec.rb:79 
- [ ] rspec ./spec/features/admin/product_import_spec.rb:130 
- [ ] rspec ./spec/features/admin/enterprises/images_spec.rb:29 
- [ ] rspec ./spec/features/admin/enterprises/images_spec.rb:66 
- [ ] rspec ./spec/features/consumer/shopping/checkout_spec.rb:149

**Comment by luisramos0** [comment]
@mkllnk I have 2 questions regarding 2 of your commits in this branch:

https://github.com/openfoodfoundation/openfoodnetwork/commit/54085be27e349f75b22354f910eee2a9aae3de76
Why is this needed? Is it fixing the spec? Maybe we can improve the commit message.

https://github.com/openfoodfoundation/openfoodnetwork/commit/8afcb5b6dd8ef79182f9ceb42f9d2b71966cd9d8
What bug is it? Is it something we should log into GH or maybe explain the expect/actual behaviour in the comment.

**Comment by mllocs** [comment]
How's the time diff of phantom vs headless chrome? In another project I'm working on we did that same migration and we noticed that headless was way slower than phantom. 

Even if tests are a bit slower it makes total sense to migrate to headless chrome, phantom is abandoned and headless chrome is way more similar to a real browser. Great job! :D

**Comment by luisramos0** [comment]
Hey Maxim!
Broken tests and retries make this build a bit slow.
But job 1 is green and running 508 specs on both master and this PR.
Master takes 10m42 and this branch is taking 6m47, we may have good news! Lets see when all jobs are green.
Master: https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/master/builds/324
This branch: https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/headless-chrome/builds/5

**Comment by Matt-Yorkley** [comment]
It's finally happening! :tada: Thanks @luisramos0

**Comment by luisramos0** [comment]
All credit goes to the ~40 commits from @oeoeaio

**Comment by mkllnk** [comment]
> 54085be
> Why is this needed? Is it fixing the spec? Maybe we can improve the commit message.

If I remember correctly, there is a Javascript error specific to embedded shopfronts that is not due to our code. It may be some security violation when trying to load Google Maps. This was the quick solution to get to a green build, but there is probably a better one.

 
> 8afcb5b#diff-2f114975ce4c238323ecb0c23daab4b7R113
> What bug is it? Is it something we should log into GH or maybe explain the expect/actual behaviour in the comment.

When you have pending changes and try to navigate away from a page, it asks you "Are you sure?". When we click the "Update" button to save changes, we need to wait until it is actually saved and "loading" disappears before doing anything else.

**Comment by luisramos0** [comment]
@mkllnk I think I'll rewrite the respective commits and add these notes as comments. ok?

**Comment by luisramos0** [comment]
actually, I am going to remove these commits and create a PR with them (including new comments). we can then decide what to do in the PR.

**Comment by luisramos0** [comment]
After #3167 we only have 4 broken tests to go.

**Comment by luisramos0** [comment]
My "in dev" here meant creating the branch and clean it up. That's done.

Now, someone needs to try to fix these 6 tests at some point.

**Comment by sigmundpetersen** [comment]
Maybe we could turn this into an Epic and create 6 issues for those? What do you think @luisramos0 ?

**Comment by luisramos0** [comment]
Maybe. Someone needs to look at the specs and decide. I dont think we have 6 problems here, maybe only 3 things to be fixed.

**Comment by sigmundpetersen** [comment]
Ping @mkllnk

**Comment by mkllnk** [comment]
The next person working on this issue can decide if this issue should be split into multiple issues or not.

**Comment by Matt-Yorkley** [comment]
I couldn't get headless to run on my machine without adding the `chromedriver-helper` gem. Shall I open a PR to add it to the gemfile?

**Comment by Matt-Yorkley** [comment]
Actually, I think the bulk_order_management will still be red. I'll take a look at it now.

**Comment by Matt-Yorkley** [comment]
If I'm not mistaken, the customers_spec test will be fixed by your refactor of the columns dropdown code @mkllnk

**Comment by luisramos0** [comment]
Amazing guys!!!
There are only 3 broken specs here:
https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/pull-request-3336/builds/2
All in bulk_order_management_spec, it looks like #3325 will make this headless-chrome green 🎉 🎉 🎉

**Comment by luisramos0** [comment]
the amazing news is that we have a green build on #3343, well done @Matt-Yorkley on the last mile!
the not so good news is that master build is taking 45min total, and this build is 57mins... it means 3 minutes slower with 4 jobs.

**Comment by Matt-Yorkley** [comment]
The timer for the individual jobs seems to be adding on waiting time to the total for each one, and there's been some queuing.

**Comment by luisramos0** [comment]
ah, yes, 48m30s in the last headless chrome!

**Comment by sauloperez** [comment]
A friend of mine told me they did exactly the same (@mllocs 's colleague) and the test suite with chrome takes longer so I'm not sure we'll win anything on that end.
